src/Streamly/Internal/Unicode/Utf8.hs

```diff
@@ -6,62 +6,160 @@
--- Type
+-- Special folds
 --------------------------------------------------------------------------------
 
--- | A space efficient, packed, unboxed Unicode container.
-newtype Utf8 =
-    Utf8 (Array Word8)
-    deriving (NFData)
+-- XXX We should write these APIs generalized on Array a and then just use those
+-- for the Utf8 type. The generalized APIs would be more useful, they can go in
+-- the Array module itself and can be used generally for arrays, you won't need
+-- to transform arrays into stream and then back for such common operations.
+-- | Concatenate a list of 'Utf8's.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE concat #-}
+concat :: [Utf8] -> Utf8
+concat ts =
+    case Prelude.filter (not . null) ts of
+        [] -> empty
+        [t] -> t
+        xs -> Prelude.foldl1 append xs
+
+-- | Map a function over a 'Utf8' that results in a 'Utf8', and
+-- concatenate the results.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE concatMap #-}
+concatMap :: (Char -> Utf8) -> Utf8 -> Utf8
+concatMap f = concat . foldr ((:) . f) []
 
 --------------------------------------------------------------------------------
--- Functions
+-- Zipping
 --------------------------------------------------------------------------------
 
-{-# INLINE toArray #-}
-toArray :: Utf8 -> Array Word8
-toArray (Utf8 arr) = arr
+-- | 'zip' takes two 'Utf8's and returns a list of
+-- corresponding pairs of bytes. If one input 'Utf8' is short,
+-- excess elements of the longer 'Utf8' are discarded. This is
+-- equivalent to a pair of 'unpack' operations.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE zip #-}
+zip :: Utf8 -> Utf8 -> [(Char,Char)]
+zip a b =
+    unsafePerformIO
+        $ Stream.toList $ Stream.zipWith (,) (toStream a) (toStream b)
 
+-- | 'zipWith' generalises 'zip' by zipping with the function
+-- given as the first argument, instead of a tupling function.
+-- Performs replacement on invalid scalar values.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE zipWith #-}
+zipWith :: (Char -> Char -> Char) -> Utf8 -> Utf8 -> Utf8
+zipWith f a b = fromStream (Stream.zipWith f (toStream a) (toStream b))
 
-{-# INLINEABLE pack #-}
-pack :: String -> Utf8
-pack s =
-    Utf8
-        $ unsafePerformIO
-        $ Array.fromStreamN len $ Unicode.encodeUtf8' $ Stream.fromList s
+-- | Breaks a 'Utf8' up into a list of words, delimited by 'Char's
+-- representing white space.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE words #-}
+words :: Utf8 -> [Utf8]
+words = split isSpace
 
-    where
+-- | Breaks a 'Utf8' up into a list of 'Utf8's at
+-- newline 'Char's. The resulting strings do not contain newlines.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE lines #-}
+lines :: Utf8 -> [Utf8]
+lines = split (== '\n')
 
-    len = length s
+-- | Joins lines, after appending a terminating newline to
+-- each.
+--
+-- /Time complexity:/ O(n)
+{-# INLINE unlines #-}
+unlines :: [Utf8] -> Utf8
+unlines = concat . List.map (`snoc` '\n')
```

Comment:
using snoc may not be most efficient, we should implement it in a streaming way, see unlines in Unicode.Stream module.